### PR TITLE
ip1 handout

### PR DESCRIPTION
**Pull request by iamabhutta**

### CODE REVIEW

File: Assignments/ip1/ip1.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+## Objectives of this assignment
+The objectives of this assignment are to:
+*  Get you familiar with the basics of TypeScript, VSCode, and the project codebase
+*  Learn how to read and write code in TypeScript
+*  Translate high-level requirements into code
+*  Learn how to write unit tests with Jest
+
+## Getting started with this assignment
+
+Before you begin, be sure to check that you have NodeJS 18.x installed, along with VSCode. We have provided a tutorial on setting up a development environment for this class.
```
Comment (JoyMitra) [inline Assignments/ip1/ip1.md]: We should probably ask them to install the LTS verison

Comment (mwand) [reply Assignments/ip1/ip1.md]: We had some discussion on this point at our last meeting. Can you confirm that IP1 runs with the LTS version 20.16.1 ?

**Comment by mwand** [comment]
The section on Submission and Grading doesn't specify that the students are supposed to run Stryker and submit the stryker report.   Have I missed something?

**Comment by mwand** [comment]
Here is my commented copy of ip1.md
[ip1 - Commented.md](https://github.com/user-attachments/files/16734762/ip1.-.Commented.md)

**Comment by mwand** [comment]
Having worked on task1 and task2 I am now very much in favor of giving the students a test suite.   The prose specifications as written leave so much to the imagination that I can't see how we could grade the submissions.

**Comment by mwand** [comment]
I am very disappointed that the codebase does not even come close to matching our coding standards [https://neu-se.github.io/CS4530-Spring-2024/policies/style/#documentation-and-comments](url).  Specifically:

```
Documentation and Comments
All public properties and methods (other than getters, setters, and constructors) must be documented using [JSDoc-style comments](https://devdocs.io/jsdoc/about-getting-started) that describes what the property/method does. Example:

/** The unique identifier for this player * */
private readonly _id: string;

/**
 * A handler to process a remote player's subscription to updates for a room
 *
 * @param socket the Socket object that we will use to communicate with the player
 */
```
I'm not doing this to be a stickler-- I really didn't understand what some of these functions were supposed to do.  The functions in models/application.ts were particular offenders in this regard.  (Speaking of which, why is this file in the 'models' directory?   It's very different from the others in that directory)

I also got confused about the naming conventions for functions:  addXX vs getXX vs fetchXX.  But that's not so important.

OK, I'm finished venting for the night.  Bye.

**Comment by iamabhutta** [comment]
Some observations from my installation experience:
1. clarify mongb installation instructions ...  will get an error if you try to connect to db thru compass. Start db first using 
"C:\Program Files\MongoDB\Server\7.0\bin\mongod.exe" --dbpath="c:\data\db"
then connect thru compass.
then run mongosh from command prompt
add instructions to create db with name "fake_so"

2. need link of repo in instructions (hw3)

3. scary looking warnings .... need to check/confirm versions needed first

4. are express, jest, and axios installed in client or server?

5. add npm start ... do we need 2 terminals?

**Comment by mwand** [comment]
Bleah!  I worked very hard to rewrite the README to make the installation smoother.   It's very confusing to have some IP1 materials in the course repo and some in the IP1 repo.   

My rewrites are in the IP1 repo, branch new-readme.  Can we untangle this, please??

File: Assignments/ip1/ip1.md [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+4. For Windows, in the Package dropdown, select `msi`.  Then download and run the installer.
+5. On Windows, select the _“Install MongoDB as a Service”_ checkbox and install.  This will start MongoDB as a background service.
+6. Install "MongoDB Compass" if prompted.
+7. Verify if the MongoDB server is running using the Windows Services app.
+   
+Mongo offers several methods of interacting with your Mongo databases.
+
+* MongoDB Compass is an interactive application for creating, modifying, and querying MongoDB connections.  It should be started as part of the installation process, showing a connection to `mongodb://localhost:27017/`.
+
+* Mongo shell (*mongosh*) provides a command-line interface that can be used to interact with databases in MongoDB. Download it [here](https://www.mongodb.com/try/download/shell) using the msi package. You can also use *mongosh* to see if the MongoDB server is running.  Try the MongoDB Community Edition and the command `show dbs`; you should see a list of existing databases in your local instance.
```
Comment (VatsalMehta27) [inline Assignments/ip1/ip1.md]: Mongo shell is automatically installed with MongoDB through the Mac installation instructions. To use it, make sure MongoDB is running as a macOS service, then type `mongosh` into the terminal.

File: Assignments/ip1/ip1.md [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+#### 1.2 Install MongoDB
+
+We use MongoDB as the NoSQL database to store data related to this application.  
+ 
+1. Follow the [instructions in the official MongoDB documentation](https://www.mongodb.com/docs/manual/administration/install-community/) to install the free community edition.
+2. Choose 'Install on Linux', 'Install on macOS', or 'Install on Windows', depending on your system.
+3. Scroll down to the section labeled 'Install MongoDB Community Edition.'and to [MongoDBDownload Center](https://www.mongodb.com/try/download/community?tck=docs_server).
+4. For Windows, in the Package dropdown, select `msi`.  Then download and run the installer.
+5. On Windows, select the _“Install MongoDB as a Service”_ checkbox and install.  This will start MongoDB as a background service.
+6. Install "MongoDB Compass" if prompted.
```
Comment (VatsalMehta27) [inline Assignments/ip1/ip1.md:48]: For Mac, download the dmg file from https://www.mongodb.com/try/download/compass. Once the application starts:
1. Click on "Add new connection" in the left sidebar.
2. Make sure the URI field is `mongodb://localhost:27017`
3. Click on "Connect" - MongoDB will need to be running

**Comment by zhang-hongyi34** [comment]
I walked through IP1 and here are my observations:

- For '3. Populate the initial database', when I run command, it gives errors because some of the properties (e.g. `description` in `tag`) is missing because it hasn't been implemented yet (which should be resolved after implementing task 2). Should we add a note to tell the students that these errors are expected?

- Currently, all files under server/controller, server/data, server/models, and some files in client have CRLF as end of line sequence, which gives a error in my editor. We can fix it by running prettier on all files

- For Task 1, I couldn't find tests for the expected behavior of `getQuestionsByFilter`, which may lead to confusion for students since they won't know if there implementation is correct. We can either add the tests, or change the task instruction to be more specific (such as stating how should the `asked_by` field be passed to `getQuestionsByFilter`

- For Task 2: 
  - We didn't mention that students need to change the schema, though I suppose this could be implied
  - It seems that `getTags` is partically implemented in the starter code, since the function already takes in an array of tag instead of an array of tag names
  - Also, `getTagByName` should be an endpoint but the instruction says it's a function
  - I also think it would be confusing to students if we don't give more instructions on how to add the `getTagByName` endpoint

- For Task 3: 
  - The second sentence of step 2 doesn't apply, since we removed the original step 1
  - I'm also curious why did we change the function name from `getMostViewedQuestion` to `sortQuestionsByMostViewed`, since now it doesn't align with other functions in `getQuestionsByOrder` (e.g. `getActiveQuestion`). If we are making this change, we should also change existing functions such as `getActiveQuestion` as well
  - The rubric also uses the old instructions. It has "Update `OrderType`" and "Add `getMostViewedQuestion`"

- For Task 4, I didn't find any issue. Though it seems to me that it's contributing too much to the overall score despite being not too much more difficult than the previous tasks

File: Assignments/ip1/ip1.md (hunk 3 above)
Comment (iamabhutta) [reply Assignments/ip1/ip1.md:48]: can u update the md file?

File: Assignments/ip1/ip1.md (hunk 2 above)
Comment (iamabhutta) [reply Assignments/ip1/ip1.md]: can u update the md file if needed?

**Pull request merged**